GetDKAN/dkan
# API page caching with cache tags

* Adds page caching with node-based cache tags to the metastore, datastore, search and SQL controllers. 
* Invalidates the distribution cache tag when a datastore import is complete

Also, cleaned up some code in a few controllers, including fully renaming the metastore controller to start to be more consistent  with controller naming, be aware of possible conflicts.

## QA Steps

(better testing steps to come)

1. Try making various requests to endpoints from the four controllers above.
2. Check that on first request to a new URL, `X-Drupal-Cache: MISS` is in the headers. On a second request, `X-Drupal-Cache: HIT` should be present, and responses should be faster.
3. Check that requests with different GET parameters are cached separately, and that cachebusters work.
4. Check `SELECT cid, tags FROM cache_page` after a request and ensure that the expected tags are set for the URL you requested.
5. Edit a dataset, and check that the expected endpoints now respond with again `X-Drupal-Cache: MISS`
6. Create a new dataset, and request its metadata from the API before running the datastore_import queue. Ensure that it is cached.
7. Run the datastore_import queue and check again. Make sure that the dataset is no longer cached.
8. Replace the file on the above dataset and perform the same check again, before and after running the import queue.

## Known issues

* We now have quite a mess of code in different places building our API responses.
* Introduced a couple new hard-coded schema names (avoided as much as possible)
